### PR TITLE
Add DoorkeeperRevisionFeedWorker for daemon to trigger HTTP requests for revision changes

### PR DESCRIPTION
**Pull request by darkwing**
## What's up

This is a super early WIP PR for adding `DoorkeeperRevisionFeedWorker`.  Here's how it works:

1.  Just the presence of this `DoorkeeperRevisionFeedWorker` class and `isEnabled` returning `true` enables it for the daemon

2.  Upon changes to a revision, the class' `publishFeedStory` is executed.

## Debugging and Testing

1.  Enter the phabricator container's shell (`docker exec -it bmoextensions_phabricator_1 /bin/sh`) and navigate to `/app/phabricator/bin`

2.  These tasks run in the `PhabricatorTaskmasterDaemon`, so execute: `./phd debug PhabricatorTaskmasterDaemon`

3.  Go to any revision, change something, save it -- the `publishFeedStory` method will execute providing the "story" information for the change

4.  Use `echo` to spit out messages from within the `publishFeedStory` function -- they will be spit out from the daemon when run!

## ProTips

1.  If you're comfortable, instead of `docker-compose down` and `docker-compose up --build` for every change, you can use `vi` to make updates to the extension code.  All you'd need to do then is take down the daemon and bring it back up.

## Output

Making an update to a revision should send an update to BMO that looks like:

```js
{
  "accepted_users": [1, 2, 3], // BMO user IDs
  "revision_id": 12, // Phab ID without "D"
  "bug_id": 123456
}
```

Obsoleting a revision should send and update to BMO that looks ike:

```js
{
  "revision_id": 12, // Phab ID without "D"
  "bug_id": 123456
}
```

### CODE REVIEW

File: doorkeeper/worker/DoorkeeperRevisionFeedWorker.php [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    private function obsolete_attachment($revision) {
+      $request_data = array(
+        'revision_id' => $revision->getID(),
+        'bug_id' => $this->get_bugzilla_bug_id($revision)
+      );
+
+      $future_uri = id(new PhutilURI(PhabricatorEnv::getEnvConfig('bugzilla.url')))
+        ->setPath('/rest/phabbugz/obsolete/');
```
Comment (dklawren) [inline doorkeeper/worker/DoorkeeperRevisionFeedWorker.php]: Remove trailing /

File: doorkeeper/worker/DoorkeeperRevisionFeedWorker.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        ->setExpectStatus(array(200));
+
+      // Debugging
+      echo pht('Making a request to: %s', (string) $future_uri);
+      echo 'Using data:'.json_encode($request_data, JSON_FORCE_OBJECT);
+
+      try {
+        list($status) = $future->resolve();
+        if($status->getStatusCode() != 200) {
+          // TODO:  What should we do in case of failure?  Re-queue?
```
Comment (dklawren) [inline doorkeeper/worker/DoorkeeperRevisionFeedWorker.php]: Does Doorkeeper have a way to re-queue built-in that doesn't require republishing to the feed?

File: doorkeeper/worker/DoorkeeperRevisionFeedWorker.php [hunk 3]
```diff
@@ -180,30 +178,31 @@ private function sendUpdateRequest($revision, $accepted_bmo_ids) {
...
+          echo(pht('obsoleteAttachment exception: %s %s', $status_code, $ex->getErrorCodeDescription($status_code)));
         }
     }
 
     private function get_http_future($uri) {
       return id(new HTTPSFuture((string) $uri))
         ->addHeader('X-Bugzilla-API-Key', PhabricatorEnv::getEnvConfig('bugzilla.automation_api_key'))
+        ->setMethod('PUT')
         ->addHeader('Accept', 'application/json')
+        ->addHeader('Content-Type', 'application/json')
```
Comment (dklawren) [inline doorkeeper/worker/DoorkeeperRevisionFeedWorker.php]: we will need to remove this as it causes the data to be passed as form-urlencoded :(

**Comment by darkwing** [comment]
Removed said line

**Pull request merged**